### PR TITLE
Upstream Timing: Time pool callback latency

### LINKED ISSUE #25480: Pending stream queue timing
Pending Stream  Queue Timing

When a stream is pending on a connection pool we should gather a timestamp. When it is not longer pending as there's stream capacity or it's cancelled we should take another timestamp and get the delta in order to record how long the stream spent in pending state.

This can help diagnose issues such as high Envoy e2e latency due to insufficient backend capacity.

### PR DESCRIPTION
**Pull request by KBaichoo**
Commit Message: Upstream Timing: Time pool callback latency.
Additional Description:
Risk Level: low
Testing: Added unit, integration test
Docs Changes: n/a
Release Notes: included
Platform Specific Features:
Fixes #25480
[Optional Fixes commit #PR or SHA]
[Optional Deprecated:]
[Optional [API Considerations](https://github.com/envoyproxy/envoy/blob/main/api/review_checklist.md):]

### CODE REVIEW

**Review by RyanTheOptimist — DISMISSED** [review]
LGTM modulo one small nit.

**Review by snowp — APPROVED** [review]
Thanks, makes sense to me modulo Ryan's comments

I remember way back I would have loved to have something like this available

File: source/common/router/upstream_request.h [hunk 1]
```diff
@@ -193,6 +193,7 @@ class UpstreamRequest : public Logger::Loggable<Logger::Id::router>,
   StreamInfo::UpstreamTiming& upstreamTiming() {
     return stream_info_.upstreamInfo()->upstreamTiming();
   }
+  void onConnectionPoolCallback();
```
[the remark this answers is not in the document]
Comment (KBaichoo) [reply source/common/router/upstream_request.h]: Good idea, done!

**Review by RyanTheOptimist — APPROVED** [review]

**Pull request merged**